docs/source/development/eep-01-pytrees.rst

```diff
@@ -0,0 +1,848 @@
+
+
+The optimization output
+-----------------------
+
+The following entries of the output of minimize are affected by the change:
+
+- ``"solution_params"``: A pytree with the same structure as ``params``
+- ``"solution_criterion"``: The output dictionary of ``crit`` evaluated solution params
+- ``solution_derivative``: Maybe we should not even have this entry.
+
+.. danger:: We need to discuss if an in which form we want to have a solution
+    derivative entry. In it's current form it is useless if constraints are used.
+    This gets worse when we allow for pytrees and translating this into a meaningful
+    shape might be very difficult.
+
+
+Add bounds
+----------
+
+Bounds on parameters that are inside a DataFrame with ``"value"`` column can simply be
+specified as before. For all others, there are separate ``lower_bounds`` and
+``upper_bounds`` arguments in ``maximize`` and ``minimize``.
+
+``lower_bounds`` and ``upper_bounds`` are pytrees of the same structure as ``params`` or
+a subtree that preserves enough structure to match all bounds. For example:
+
+.. code-block:: python
+
+    minimize(
+        criterion=crit,
+        params=params,
+        algorithm="scipy_lbfgsb",
+        lower_bounds={"delta": 0},
+        upper_bounds={"delta": 1},
+    )
+
+This would add bounds for delta, keep the bounds on all ``"utility"`` parameters, and
+leave the ``"probs"`` parameters unbounded.
+
+
+Add a constraint
+----------------
+
+Currently, parameters to which a constraint is applied are selected via a ``"loc"`` or
+``"query"`` entry in the constraints dictionary.
+
+This keeps working as long as params are specified as a single DataFrame containing a
+``"value"`` column. If a more general pytree is used we need a "selector" entry instead.
+The value of that entry is a callable that takes the pytree and returns selected
+parameters.
+
+The selected parameters can be returned as estimagic-pytrees. For constraints where
+order plays a role (e.g., increasing), the order defined by ``tree_flatten`` is used.
+However we advise users to return one-dimensional arrays in that case, so they do not
+have to learn the internal of ``tree_flatten``.
+
+As an example, let's add probability constraints for each row of ``"probs"``:
+
+.. code-block:: python
+
+    constraints = [
+        {"selector": lambda params: params["probs"][0], "type": "probability"},
+        {"selector": lambda params: params["probs"][1], "type": "probability"},
+    ]
+
+    minimize(
+        criterion=crit,
+        params=params,
+        algorithm="scipy_lbfgsb",
+        constraints=constraints,
+    )
+
+Internally, constraints are already applied on a 1 dimensional numpy array and the
+parameter selections specified by "loc" and "query" are translated into positions in
+that array. What changes is that we now also have to translate the parameter selections
+from "selector" functions into positions. This is trivial by calling ``tree_unflatten``
```

Comment:
Maybe too much detail on implementation, likely unclear for anyone but @janosg. Either drop or expand.

Reply:
Replaced it by a shorter and more abstract description that mainly just states that the implementation will not be too hard.